lib/semantics/check-do.cc

```diff
@@ -647,20 +643,124 @@ void DoChecker::Leave(const parser::DoConstruct &x) {
+}
+
+// Check that CYCLE and EXIT statements do not cause flow of control to
+// leave DO CONCURRENT, CRITICAL, or CHANGE TEAM constructs.
+void DoChecker::CheckForBadLeave(
+    StmtType stmtType, const ConstructNode &construct) const {
+  std::visit(
+      common::visitors{
+          [&](const parser::DoConstruct *doConstructPtr) {
+            if (doConstructPtr->IsDoConcurrent()) {
+              // C1135 and C1167
+              SayBadLeave(stmtType, "DO CONCURRENT", construct);
+            }
+          },
+          [&](const parser::CriticalConstruct *) {
+            // C1135 and C1168
+            SayBadLeave(stmtType, "CRITICAL", construct);
+          },
+          [&](const parser::ChangeTeamConstruct *) {
+            // C1135 and C1168
+            SayBadLeave(stmtType, "CHANGE TEAM", construct);
+          },
+          [](const auto *) {},
+      },
+      construct);
+}
+
+static bool isExit(StmtType stmtType) { return stmtType == StmtType::EXIT; }
+
+static bool StmtMatchesConstruct(const parser::Name *stmtName,
+    StmtType stmtType, const parser::Name *constructName,
+    const ConstructNode &construct) {
+  bool inDoConstruct{MaybeGetDoConstruct(construct) != nullptr};
+  if (stmtName == nullptr) {
+    return inDoConstruct;  // Unlabeled statements match all DO constructs
+  } else if (constructName && constructName->source == stmtName->source) {
+    return isExit(stmtType) || inDoConstruct;
+  } else {
+    return false;
   }
 }
 
-// C1166
-void DoChecker::Enter(const parser::ExitStmt &) {
-  if (!context_.InsideDoConstruct()) {
-    context_.Say(
-        *context_.location(), "EXIT must be within a DO construct"_err_en_US);
+// C1167 Can't EXIT from a DO CONCURRENT
+void DoChecker::CheckDoConcurrentExit(
+    StmtType stmtType, const ConstructNode &construct) const {
+  if (isExit(stmtType) && ConstructIsDoConcurrent(construct)) {
+    SayBadLeave(StmtType::EXIT, "DO CONCURRENT", construct);
   }
 }
 
+// Check nesting violations for a CYCLE or EXIT statement.  Loop up the nesting
+// levels looking for a construct that matches the CYCLE or EXIT statment.  At
+// every construct, check for a violation.  If we find a match without finding
+// a violation, the check is complete.
+void DoChecker::CheckNesting(
+    StmtType stmtType, const parser::Name *stmtName) const {
+  const ConstructStack &stack{context_.constructStack()};
+  for (auto rIter{stack.crbegin()}; rIter != stack.crend(); ++rIter) {
+    const ConstructNode &construct{*rIter};
+    const parser::Name *constructName{MaybeGetNodeName(construct)};
+    if (StmtMatchesConstruct(stmtName, stmtType, constructName, construct)) {
+      CheckDoConcurrentExit(stmtType, construct);
+      return;  // We got a match, so we're finished checking
+    }
+    CheckForBadLeave(stmtType, construct);
+  }
+
+  // We haven't found a match in the enclosing constructs
+  if (isExit(stmtType)) {
+    context_.Say("No matching construct for EXIT statement"_err_en_US);
```

Comment:
```suggestion
    context_.Say("No matching DO construct for %s statement"_err_en_US, EnumToString(stmtType));
```

This will avoid the if-else block. Assuming the `DO` should be there in both messages.

Reply:
If the `DO` was omitted on purposed disregard my comment.

Reply:
Yes, ```CYCLE``` statements must specifically match a ```DO``` construct, whereas ```EXIT``` statements can match a variety of constructs.